src/queue/priority.rs

```diff
@@ -70,11 +70,39 @@ where
     }
 }
 
+/// Priority represents the priority of a task.
+/// Tasks with smaller priority value will be scheduled earlier than bigger ones.
+#[derive(PartialEq, Copy, Clone, Eq, PartialOrd)]
+pub struct Priority {
+    /// The priority of the group this task belongs to, ordered in descending order.
+    pub group_priority: u64,
+    /// The priority value of this task.
+    pub value: u64,
+}
+
+impl Default for Priority {
+    fn default() -> Self {
+        Priority {
+            group_priority: 8,
+            value: 0,
+        }
+    }
+}
+
+impl Ord for Priority {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        other
+            .group_priority
+            .cmp(&self.group_priority)
+            .then_with(|| self.value.cmp(&other.value))
+    }
+}
+
 /// A trait used to generate priority value for each task.
 pub trait TaskPriorityProvider: Send + Sync + 'static {
     /// Return a priority value of this task, all tasks in the priority
     /// queue is ordered by this value.
-    fn priority_of(&self, extras: &Extras) -> u64;
+    fn priority_of(&self, extras: &Extras) -> Priority;
```

Comment:
Typically, the total number of priorities is not very big, so what about sharing the high bits of the value field, so we can still use a u64 here and keep this trait simple.

Reply:
I don't think it makes the trait complicated.  A defined struct is more readable.

Reply:
But not all case need the `task_priority`, this change some what put some specified details into this trait